source/tricky-pipe/src/tp3.rs

```diff
@@ -0,0 +1,910 @@
+    }
+
+    /// Returns `true` if this channel is full.
+    ///
+    /// If this method returns `true`, then any calls to [`Sender::reserve`] or
+    /// [`SerSender::reserve`] will yield until the queue is empty. Any calls to
+    /// [`Sender::try_reserve`] or [`SerSender::try_send`] will return an error.
+    #[inline]
+    #[must_use]
+    pub fn is_full(&self) -> bool {
+        self.pipe.core().is_full()
+    }
+
+    /// Returns the number of messages currently in the channel.
+    #[inline]
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.pipe.core().len()
+    }
+
+    /// Returns the **maximum capacity** of the channel.
+    ///
+    /// This is the maximum number of messages that may be queued before senders
+    /// must wait for additional capacity to become available. The capacity of
+    /// the channel is determined *when it is constructed*, and the value
+    /// returned by this method will never change over the channel's lifetime,
+    /// regardless of the current [length](Self::len) of the channel.
+    ///
+    /// To determine the current remaining capacity in the channel, use the
+    /// [`remaining`](Self::remaining) method, instead.
+    #[inline]
+    #[must_use]
+    pub fn capacity(&self) -> usize {
+        self.pipe.core().capacity as usize
+    }
+
+    /// Returns the **current remaining capacity** of this channel.
+    ///
+    /// This is equivalent to subtracting [`self.len()`](Self::len) from
+    /// [`self.capacity()`](Self::capacity).
+    #[inline]
+    #[must_use]
+    pub fn remaining(&self) -> usize {
+        self.len() - self.capacity()
+    }
+}
+
+impl<T: 'static> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        self.pipe.core().add_tx();
+        Self {
+            pipe: self.pipe.clone(),
+        }
+    }
+}
+
+impl<T: 'static> Drop for Sender<T> {
+    fn drop(&mut self) {
+        self.pipe.core().drop_tx();
+    }
+}
+
+impl<T> Permit<'_, T> {
+    pub fn send(self, val: T) {
+        // write the value...
+        unsafe {
+            // safety: because we allocated the slot's index, we have exclusive
+            // mutable access to this slot.
+            self.cell.deref().write(val);
+        }
+        // ...and commit.
+        self.pipe.commit_send();
+    }
+
+    /// Send the current value of the reserved slot to the channel.
+    ///
+    /// This method is intended to be used in conjunction with the
+    /// [`DerefMut`]`<Target = `[`MaybeUninit`]`<T>>` implementation for
+    /// `Permit` to allow writing to the reserved slot in the channel's buffer
+    /// in place, rather than by moving a value into the slot. This may, in some
+    /// cases, be more efficient when the messages are large.
+    ///
+    /// # Safety
+    ///
+    /// Calling `commit` without writing to the reserved slot **will result in
+    /// the [`Receiver`] reading uninitialized memory**! Ensure that the slot
+    /// has been initialized prior to calling this method!
```

Comment:
we could also just not have this API (and get rid of the `Deref`/`DerefMut` impls...i'm not sure if this is actually useful enough to justify a potential "send the receiver some uninitialized memory lol" footgun...